airflow/models/base_async_operator.py

```diff
@@ -0,0 +1,96 @@
+# -*- coding: utf-8 -*-
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+"""
+Base Asynchronous Operator for kicking off a long running
+operations and polling for completion with reschedule mode.
+"""
+from functools import wraps
+from airflow.sensors.base_sensor_operator import BaseSensorOperator
+from airflow.exceptions import AirflowException
+from airflow.models.xcom import XCOM_EXTERNAL_RESOURCE_ID_KEY
+
+class BaseAsyncOperator(BaseSensorOperator, SkipMixin):
+    """
+    AsyncOperators are derived from this class and inherit these attributes.
+
+    AsyncOperators must define a `submit_request` to fire a request for a
+    long running operation with a method and then executes a `poke` method
+    executing at a time interval and succeed when a criteria is met and fail
+    if and when they time out. They are effctively an opinionated way use
+    combine an Operator and a Sensor in order to kick off a long running
+    process without blocking a worker slot while waiting for the long running
+    process to complete by leveraging reschedule mode.
+
+    :param soft_fail: Set to true to mark the task as SKIPPED on failure
+    :type soft_fail: bool
+    :param poke_interval: Time in seconds that the job should wait in
+        between each tries
+    :type poke_interval: int
+    :param timeout: Time, in seconds before the task times out and fails.
+    :type timeout: int
+    :type mode: str
+    """
+    ui_color = '#9933ff'  # type: str
+    valid_modes = ['poke', 'reschedule']  # type: Iterable[str]
+
+    @apply_defaults
+    def __init__(self,
+                 *args,
+                 **kwargs) -> None:
+        super().__init__(mode='reschedule', *args, **kwargs)
+
+    def submit_request(self, context) -> string:
```

Comment:
```suggestion
    def submit_request(self, context) -> IT:
```
I imagine that the identifier will be a different type, e.g. a number, or an array with several identifiers. What do you think about using [generic types](https://docs.python.org/3.5/library/typing.html#typing.Generic)?

Reply:
@mik-laj why not string / dict / json-serializable string?  if this is using xcom, and xcom pickling is going away, is that not a good solution?  Even if we were to add a new database column and not use xcom, wouldn't json serializability be desired?  (have not really used generics and not sure what their value is here... interested in understanding more about what you are suggesting)

Reply:
The generic class allows you to provide information about the type, and at the same time specify that the same type occurs in several places. It can still be string or dict or other serializable value, but it will not be specified by this class, but it will have to be determined by the child class.

For example: We have a Stack class. The list can contain two methods: `push(A) -> None`, `poll() -> A`;  If we want to point out that the given A in both cases must be equal in the child class, we can mark it using generic. In this case, the child class will have to specify the type, but we won't have to do it because we don't know what type it can be. I know it should be a subset of types, but I don't know what type exactly.